### PR TITLE
feat(generator): generate delete methods in containers

### PR DESCRIPTION
**Pull request by archerzz**
# Description

- update `ResourceContainerWriter` to add delete methods generation
- update test cases accordingly
- refactor to move HTTP method type check logic into `ClientMethod` and `RestClientMethod`

AB#5998

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Output/Models/ClientMethod.cs [hunk 1]
```diff
@@ -21,5 +21,21 @@ public ClientMethod(string name, RestClientMethod restClientMethod, string? desc
...
         public Diagnostic Diagnostics { get; }
         public string Accessibility { get; }
+
+        public bool IsPut
+        {
+            get
+            {
+                return this.RestClientMethod.IsPut;
+            }
+        }
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Output/Models/ClientMethod.cs:31]: We could simplify this 
```suggestion
        public bool IsPut => this.RestClientMethod.IsPut;
```

File: test/AutoRest.TestServer.Tests/lro.cs [hunk 2]
```diff
@@ -101,20 +101,6 @@ public class LroTest: TestServerTestBase
...
-            var operation = await new LROsClient(ClientDiagnostics, pipeline, host).StartPatch200SucceededIgnoreHeadersAsync();
-            var resp = await operation.WaitForCompletionAsync().ConfigureAwait(false);
-        });
-
-        [Test]
-        public Task LROPatchInlineCompleteIgnoreHeaders_Sync() => Test((host, pipeline) =>
-        {
-            var operation = new LROsClient(ClientDiagnostics, pipeline, host).StartPatch200SucceededIgnoreHeaders();
-            var resp = WaitForCompletion(operation);
-        });
```
Comment (ArcturusZhang) [inline test/AutoRest.TestServer.Tests/lro.cs:116]: Why are we removing these tests?

**Review by ArcturusZhang — COMMENTED** [review]
Since the version of `autorest.testserver` has been updated recently, please ensure to run `npm install` first before you regenerate all the test projects

File: test/TestServerProjects/lro/Generated/LROsClient.cs [hunk 3]
```diff
@@ -74,44 +74,6 @@ public virtual LROsPut200SucceededOperation StartPut200Succeeded(Product product
...
-                var originalResponse = await RestClient.Patch200SucceededIgnoreHeadersAsync(product, cancellationToken).ConfigureAwait(false);
-                return new LROsPatch200SucceededIgnoreHeadersOperation(_clientDiagnostics, _pipeline, RestClient.CreatePatch200SucceededIgnoreHeadersRequest(product).Request, originalResponse);
-            }
-            catch (Exception e)
-            {
-                scope.Failed(e);
-                throw;
-            }
-        }
-
```
Comment (ShivangiReja) [inline test/TestServerProjects/lro/Generated/LROsClient.cs:95]: Test server version has updated. I think you need to run `npm install`, after that you will not see these changes.

File: samples/Azure.ResourceManager.Sample/Generated/AvailabilitySetContainer.cs [hunk 4]
```diff
@@ -437,6 +437,106 @@ public virtual AsyncPageable<GenericResource> GetAllAsGenericResourcesAsync(stri
...
+            {
+                var operation = await StartDeleteAsync(availabilitySetName, cancellationToken).ConfigureAwait(false);
+                return await operation.WaitForCompletionResponseAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
```
Comment (ShivangiReja) [inline samples/Azure.ResourceManager.Sample/Generated/AvailabilitySetContainer.cs:463]: Why are we generating one delete method in two classes([Resource class](https://github.com/Azure/autorest.csharp/blob/456c9e62c2a7519862473468abeedbc5189e6e08/samples/Azure.ResourceManager.Sample/Generated/AvailabilitySet.cs#L131-L165) and ResourceContainer class)?

File: test/AutoRest.TestServer.Tests/lro.cs (hunk 2 above)
Comment (archerzz) [reply test/AutoRest.TestServer.Tests/lro.cs:116]: Thanks. I'm about to ask why the generated codes are so different :)

File: samples/Azure.ResourceManager.Sample/Generated/AvailabilitySetContainer.cs (hunk 4 above)
Comment (archerzz) [reply samples/Azure.ResourceManager.Sample/Generated/AvailabilitySetContainer.cs:463]: Yes. That's what discussed in [ADO 5998](). We found that if users want to delete a resource by name, they should be able to delete without getting the resource first. Adding `Delete` to the container will save one unnecessary HTTP Get request in some cases.

This is not a preview blocker. I'll hold on this until other changes for preview gets in first.

File: src/AutoRest.CSharp/Common/Output/Models/ClientMethod.cs [hunk 5]
```diff
@@ -21,5 +21,9 @@ public ClientMethod(string name, RestClientMethod restClientMethod, string? desc
         public string? Description { get; }
         public Diagnostic Diagnostics { get; }
         public string Accessibility { get; }
+
+        public bool IsPut => this.RestClientMethod.IsPut;
```
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Output/Models/ClientMethod.cs:25]: Please remove these properties from the `ClientMethod` type. This type describes the features of the generated method, not its underlying request or requests

**Comment by archerzz** [comment]
After request path based generation change, this PR is no longer valid. I'll create a new one for this.

**Comment by archerzz** [comment]
new PR is opened: https://github.com/Azure/autorest.csharp/pull/1813

**Pull request closed without merging**